packages/sdk-metrics/src/export/PeriodicExportingMetricReader.ts

```diff
@@ -117,11 +118,24 @@ export class PeriodicExportingMetricReader extends MetricReader {
       );
     }
 
-    const result = await internal._export(this._exporter, resourceMetrics);
-    if (result.code !== ExportResultCode.SUCCESS) {
-      throw new Error(
-        `PeriodicExportingMetricReader: metrics export failed (error ${result.error})`
-      );
+    const doExport = async () => {
+      const result = await internal._export(this._exporter, resourceMetrics);
+      if (result.code !== ExportResultCode.SUCCESS) {
+        throw new Error(
+          `PeriodicExportingMetricReader: metrics export failed (error ${result.error})`
+        );
+      }
+    };
+
+    // Avoid scheduling a promise to make the behavior more predictable and easier to test
+    if (resourceMetrics.resource.asyncAttributesPending) {
+      resourceMetrics.resource
+        .waitForAsyncAttributes?.()
+        .then(doExport, err =>
```

Comment:
If `waitForAsyncAttributes` is `undefined` (old implementation) than `.waitForAsyncAttributes?.()` will resolve to `undefined` which will throw when `.then` is invoked.

Are there any unit tests for this case?

Reply:
this block is in `if (resourceMetrics.resource.asyncAttributesPending)` so when using the old implementation it will not reach that part of the code.

Reply:
right :)